File: utils/browser_cookies.py
"""
从主流浏览器获取 GitHub Cookie 的工具模块
支持：Chrome、Edge、Firefox
平台：Windows、Linux (Debian/Ubuntu)、macOS
"""

import os
import sys
import sqlite3
import json
import hashlib
from typing import Optional, Dict, List
from pathlib import Path


class BrowserCookieExtractor:
    """浏览器 Cookie 提取器"""
    
    # 各浏览器在不同平台的 Cookie 路径
    BROWSER_PATHS = {
        'chrome': {
            'win': r'%LOCALAPPDATA%\Google\Chrome\User Data\Default\Cookies',
            'linux': '~/.config/google-chrome/Default/Cookies',
            'debian': '~/.config/google-chrome/Default/Cookies',
            'mac': '~/Library/Application Support/Google/Chrome/Default/Cookies'
        },
        'edge': {
            'win': r'%LOCALAPPDATA%\Microsoft\Edge\User Data\Default\Cookies',
            'linux': '~/.config/microsoft-edge/Default/Cookies',
            'debian': '~/.config/microsoft-edge/Default/Cookies',
            'mac': '~/Library/Application Support/Microsoft Edge/Default/Cookies'
        },
        'firefox': {
            'win': r'%APPDATA%\Mozilla\Firefox\Profiles\\',
            'linux': '~/.mozilla/firefox/',
            'debian': '~/.mozilla/firefox/',
            'mac': '~/Library/Application Support/Firefox/Profiles/'
        }
    }
    
    # Linux 下检测 Debian/Ubuntu 的特征文件
    DEBIAN_MARKER = '/etc/debian_version'
    UBUNTU_MARKER = '/etc/lsb-release'
    
    def __init__(self):
        self.platform = self._detect_platform()
        self.is_debian = self._is_debian_based()
    
    def _detect_platform(self) -> str:
        """检测操作系统平台"""
        if os.name == 'nt':
            return 'win'
        elif os.name == 'posix':
            if sys.platform == 'darwin':
                return 'mac'
            return 'linux'
        return 'linux'
    
    def _is_debian_based(self) -> bool:
        """检测是否为 Debian/Ubuntu 系系统"""
        return os.path.exists(self.DEBIAN_MARKER) or os.path.exists(self.UBUNTU_MARKER)
    
    def _get_browser_path(self, browser: str) -> Optional[str]:
        """获取浏览器 Cookie 文件路径"""
        paths = self.BROWSER_PATHS.get(browser, {})
        
        # 优先尝试平台特定路径
        path_key = self.platform
        path = paths.get(path_key)
        
        if path:
            path = os.path.expandvars(os.path.expanduser(path))
            if os.path.exists(path):
                return path
        
        # Linux 下尝试 Debian 特定路径
        if self.platform == 'linux':
            debian_path = paths.get('debian')
            if debian_path:
                debian_path = os.path.expandvars(os.path.expanduser(debian_path))
                if os.path.exists(debian_path):
                    return debian_path
        
        # 兜底：尝试 linux 路径
        linux_path = paths.get('linux')
        if linux_path:
            linux_path = os.path.expandvars(os.path.expanduser(linux_path))
            if os.path.exists(linux_path):
                return linux_path
        
        return None
    
    def get_github_cookie_from_chrome(self) -> Optional[Dict]:
        """从 Chrome 获取 GitHub Cookie"""
        return self._extract_chromium_cookie('chrome')
    
    def get_github_cookie_from_edge(self) -> Optional[Dict]:
        """从 Edge 获取 GitHub Cookie（支持 Debian/Ubuntu）"""
        return self._extract_chromium_cookie('edge')
    
    def get_github_cookie_from_firefox(self) -> Optional[Dict]:
        """从 Firefox 获取 GitHub Cookie"""
        return self._extract_firefox_cookie()
    
    def _extract_chromium_cookie(self, browser: str, domain: str = 'github.com') -> Optional[Dict]:
        """
        从 Chromium 系浏览器（Chrome/Edge）提取 Cookie
        支持 Windows、Linux (Debian/Ubuntu)、macOS
        """
        cookie_path = self._get_browser_path(browser)
        
        if not cookie_path:
            return None
        
        if not os.path.exists(cookie_path):
            return None
        
        # 检查文件是否可读
        if not os.access(cookie_path, os.R_OK):
            return None
        
        cookies = {}
        conn = None
        
        try:
            # 复制数据库到临时位置（避免锁定问题）
            import tempfile
            import shutil
            
            temp_db = tempfile.NamedTemporaryFile(delete=False, suffix='.db')
            temp_path = temp_db.name
            temp_db.close()
            
            # 复制 Cookie 数据库
            shutil.copy2(cookie_path, temp_path)
            
            # 连接临时数据库
            conn = sqlite3.connect(temp_path)
            cursor = conn.cursor()
            
            # 查询 GitHub 相关 Cookie
            cursor.execute(
                'SELECT name, encrypted_value FROM cookies WHERE host_key LIKE ?',
                (f'%{domain}%',)
            )
            
            for name, encrypted_value in cursor.fetchall():
                if encrypted_value:
                    try:
                        decrypted = self._decrypt_chromium_value(encrypted_value, browser)
                        if decrypted:
                            cookies[name] = decrypted
                    except Exception:
                        # 解密失败，尝试直接使用原始值
                        try:
                            cookies[name] = encrypted_value.decode('utf-8', errors='ignore')
                        except Exception:
                            continue
            
            if conn:
                conn.close()
            conn = None
            
            # 清理临时文件
            os.unlink(temp_path)
            
        except sqlite3.Error:
            if conn:
                try:
                    conn.close()
                except Exception:
                    pass
            return None
        except Exception:
            return None
        finally:
            if conn:
                try:
                    conn.close()
                except Exception:
                    pass
        
        return cookies if cookies else None
    
    def _decrypt_chromium_value(self, encrypted_value: bytes, browser: str) -> Optional[str]:
        """
        解密 Chromium Cookie 值
        Windows: 使用 DPAPI
        Linux/macOS: 使用 AES 加密
        """
        try:
            # v10/v11 开头表示 AES 加密
            if isinstance(encrypted_value, bytes) and (encrypted_value.startswith(b'v10') or encrypted_value.startswith(b'v11')):
                return self._decrypt_aes_cookie(encrypted_value, browser)
            
            # 已解密的字符串
            if isinstance(encrypted_value, str):
                return encrypted_value
            
            # 尝试直接解码
            return encrypted_value.decode('utf-8', errors='ignore')
        except Exception:
            return None
    
    def _decrypt_aes_cookie(self, encrypted_value: bytes, browser: str) -> Optional[str]:
        """
        使用 AES 解密 Cookie
        Linux/macOS: 从密钥环获取密码
        """
        try:
            # 提取加密数据部分
            nonce = encrypted_value[3:15]  # 12 字节
            ciphertext = encrypted_value[15:-16]  # 密文
            tag = encrypted_value[-16:]  # 16 字节认证标签
            
            # 获取密钥
            key = self._get_linux_key(browser)
            if not key:
                return None
            
            # AES GCM 解密
            from Crypto.Cipher import AES
            cipher = AES.new(key, AES.MODE_GCM, nonce=nonce)
            decrypted = cipher.decrypt_and_verify(ciphertext, tag)
            
            return decrypted.decode('utf-8')
        except Exception:
            return None
    
    def _get_linux_key(self, browser: str) -> Optional[bytes]:
        """
        获取 Linux 下的解密密钥
        从密钥环或配置文件读取
        """
        key = None
        
        # 尝试从密钥环获取
        try:
            import keyring
            service_name = f"{browser.capitalize()} Safe Storage"
            key = keyring.get_password(service_name, service_name)
            
            if key:
                return self._derive_key(key)
        except Exception:
            pass
        
        # 备用：尝试从本地状态文件读取
        local_state_path = self._get_local_state_path(browser)
        if local_state_path and os.path.exists(local_state_path):
            try:
                with open(local_state_path, 'r', encoding='utf-8') as f:
                    local_state = json.load(f)
                
                key = local_state.get('os_crypt', {}).get('encrypted_key')
                if key:
                    import base64
                    key = base64.b64decode(key)
                    if key.startswith(b'DPAPI'):
                        key = key[5:]
                    return None
            except Exception:
                pass
        
        # 兜底：使用默认密钥
        return self._derive_key(browser + ' Safe Storage')
    
    def _derive_key(self, password: str) -> bytes:
        """派生 AES 密钥"""
        return hashlib.pbkdf2_hmac('sha1', password.encode('utf-8'), b'saltysalt', 1, 16)
    
    def _get_local_state_path(self, browser: str) -> Optional[str]:
        """获取浏览器的 Local State 文件路径"""
        browser_base_paths = {
            'chrome': '~/.config/google-chrome/',
            'edge': '~/.config/microsoft-edge/',
            'chromium': '~/.config/chromium/'
        }
        
        base_path = browser_base_paths.get(browser)
        if base_path:
            return os.path.expanduser(os.path.join(base_path, 'Local State'))
        
        return None
    
    def _extract_firefox_cookie(self, domain: str = 'github.com') -> Optional[Dict]:
        """
        从 Firefox 提取 Cookie
        Firefox Cookie 未加密，直接读取 SQLite
        """
        profiles_path = self._get_browser_path('firefox')
        
        if not profiles_path or not os.path.exists(profiles_path):
            return None
        
        # 查找 profiles.ini
        profiles_ini = os.path.join(os.path.dirname(profiles_path.rstrip('/')), 'profiles.ini')
        profiles = []
        
        if os.path.exists(profiles_ini):
            profiles = self._parse_firefox_profiles(profiles_ini, profiles_path)
        else:
            # 直接扫描目录
            try:
                for item in os.listdir(profiles_path):
                    if item.endswith('.default') or item.endswith('.default-release'):
                        profile_path = os.path.join(profiles_path, item)
                        if os.path.isdir(profile_path):
                            profiles.append(profile_path)
            except Exception:
                pass
        
        # 尝试每个 profile 的 Cookie 数据库
        for profile in profiles:
            cookies_db = os.path.join(profile, 'cookies.sqlite')
            if os.path.exists(cookies_db):
                cookies = self._read_firefox_cookies(cookies_db, domain)
                if cookies:
                    return cookies
        
        return None
    
    def _parse_firefox_profiles(self, profiles_ini: str, base_path: str) -> List[str]:
        """解析 Firefox profiles.ini 文件"""
        profiles = []
        
        try:
            import configparser
            config = configparser.ConfigParser()
            config.read(profiles_ini, encoding='utf-8')
            
            for section in config.sections():
                if section.startswith('Profile'):
                    path = config.get(section, 'Path', fallback=None)
                    if path:
                        is_relative = config.getboolean(section, 'IsRelative', fallback=False)
                        if is_relative:
                            full_path = os.path.join(base_path, path)
                        else:
                            full_path = path
                        
                        if os.path.isdir(full_path):
                            profiles.append(full_path)
        except Exception:
            pass
        
        return profiles
    
    def _read_firefox_cookies(self, db_path: str, domain: str) -> Optional[Dict]:
        """读取 Firefox Cookie 数据库"""
        cookies = {}
        conn = None
        
        try:
            import tempfile
            import shutil
            
            temp_db = tempfile.NamedTemporaryFile(delete=False, suffix='.sqlite')
            temp_path = temp_db.name
            temp_db.close()
            
            shutil.copy2(db_path, temp_path)
            
            conn = sqlite3.connect(temp_path)
            cursor = conn.cursor()
            
            cursor.execute(
                'SELECT name, value FROM moz_cookies WHERE host LIKE ?',
                (f'%{domain}%',)
            )
            
            for name, value in cursor.fetchall():
                cookies[name] = value
            
            if conn:
                conn.close()
            conn = None
            os.unlink(temp_path)
            
        except Exception:
            if conn:
                try:
                    conn.close()
                except Exception:
                    pass
            return None
        
        return cookies if cookies else None
    
    def get_all_github_cookies(self) -> Dict[str, Optional[Dict]]:
        """获取所有浏览器的 GitHub Cookie"""
        results = {}
        
        chrome = self.get_github_cookie_from_chrome()
        if chrome:
            results['chrome'] = chrome
        
        edge = self.get_github_cookie_from_edge()
        if edge:
            results['edge'] = edge
        
        firefox = self.get_github_cookie_from_firefox()
        if firefox:
            results['firefox'] = firefox
        
        return results
    
    def list_available_browsers(self) -> List[str]:
        """列出当前系统上可用的浏览器"""
        available = []
        
        for browser in ['chrome', 'edge', 'firefox']:
            path = self._get_browser_path(browser)
            if path and os.path.exists(path):
                available.append(browser)
        
        return available
    
    def format_cookie_string(self, cookies: Dict[str, str]) -> str:
        """将 Cookie 字典格式化为字符串"""
        return '; '.join(f'{k}={v}' for k, v in cookies.items())


# 便捷函数

def get_github_cookie_from_browser(browser: str = 'auto') -> Optional[Dict]:
    """从指定浏览器获取 GitHub Cookie"""
    extractor = BrowserCookieExtractor()
    
    if browser == 'auto':
        all_cookies = extractor.get_all_github_cookies()
        if all_cookies:
            return list(all_cookies.values())[0]
    else:
        method_map = {
            'chrome': extractor.get_github_cookie_from_chrome,
            'edge': extractor.get_github_cookie_from_edge,
            'firefox': extractor.get_github_cookie_from_firefox
        }
        
        if browser in method_map:
            return method_map[browser]()
    
    return None


def get_github_cookie_string(browser: str = 'auto') -> Optional[str]:
    """获取 GitHub Cookie 字符串"""
    extractor = BrowserCookieExtractor()
    cookies = get_github_cookie_from_browser(browser)
    
    if cookies:
        return extractor.format_cookie_string(cookies)
    
    return None


def list_available_browsers() -> List[str]:
    """返回当前系统上可用的浏览器列表"""
    extractor = BrowserCookieExtractor()
    return extractor.list_available_browsers()

File: utils/browser_cookies.cpp
#include "utils/browser_cookies.h"

#include <QFile>
#include <QFileInfo>
#include <QDir>
#include <QJsonDocument>
#include <QJsonObject>
#include <QTextStream>
#include <QStandardPaths>
#include <QTemporaryFile>

#include <sqlite3.h>
#include <openssl/evp.h>
#include <openssl/sha.h>
#include <openssl/rand.h>

BrowserCookieExtractor::BrowserCookieExtractor()
    : m_platform(detectPlatform())
    , m_isDebian(detectIsDebian())
{
}

QString BrowserCookieExtractor::detectPlatform()
{
#ifdef Q_OS_WIN
    return "win";
#elif defined(Q_OS_MACOS)
    return "mac";
#else
    return "linux";
#endif
}

bool BrowserCookieExtractor::detectIsDebian()
{
    return QFileInfo::exists("/etc/debian_version") || QFileInfo::exists("/etc/lsb-release");
}

QString BrowserCookieExtractor::getBrowserPath(const QString &browser)
{
    if (browser == "chrome") {
        if (m_platform == "win")
            return QDir::fromNativeSeparators(qgetenv("LOCALAPPDATA"))
                   + "/Google/Chrome/User Data/Default/Cookies";
        else if (m_platform == "mac")
            return QDir::homePath() + "/Library/Application Support/Google/Chrome/Default/Cookies";
        else
            return QDir::homePath() + "/.config/google-chrome/Default/Cookies";
    } else if (browser == "edge") {
        if (m_platform == "win")
            return QDir::fromNativeSeparators(qgetenv("LOCALAPPDATA"))
                   + "/Microsoft/Edge/User Data/Default/Cookies";
        else if (m_platform == "mac")
            return QDir::homePath() + "/Library/Application Support/Microsoft Edge/Default/Cookies";
        else
            return QDir::homePath() + "/.config/microsoft-edge/Default/Cookies";
    } else if (browser == "firefox") {
        if (m_platform == "win")
            return QDir::fromNativeSeparators(qgetenv("APPDATA")) + "/Mozilla/Firefox/Profiles/";
        else if (m_platform == "mac")
            return QDir::homePath() + "/Library/Application Support/Firefox/Profiles/";
        else
            return QDir::homePath() + "/.mozilla/firefox/";
    }
    return QString();
}

QMap<QString, QString> BrowserCookieExtractor::getGithubCookieFromChrome()
{
    return extractChromiumCookie("chrome");
}

QMap<QString, QString> BrowserCookieExtractor::getGithubCookieFromEdge()
{
    return extractChromiumCookie("edge");
}

QMap<QString, QString> BrowserCookieExtractor::getGithubCookieFromFirefox()
{
    return extractFirefoxCookie();
}

QMap<QString, QString> BrowserCookieExtractor::extractChromiumCookie(const QString &browser, const QString &domain)
{
    QString cookiePath = getBrowserPath(browser);
    if (cookiePath.isEmpty() || !QFileInfo::exists(cookiePath))
        return {};

    QFileInfo fi(cookiePath);
    if (!fi.isReadable())
        return {};

    QMap<QString, QString> cookies;

    QTemporaryFile tmpFile;
    if (!tmpFile.open())
        return {};
    QString tmpPath = tmpFile.fileName();
    tmpFile.close();

    if (!QFile::copy(cookiePath, tmpPath))
        return {};

    sqlite3 *db = nullptr;
    if (sqlite3_open(tmpPath.toUtf8().constData(), &db) != SQLITE_OK) {
        if (db) sqlite3_close(db);
        QFile::remove(tmpPath);
        return {};
    }

    QString sql = "SELECT name, encrypted_value FROM cookies WHERE host_key LIKE ?";
    sqlite3_stmt *stmt = nullptr;
    QString pattern = "%" + domain + "%";

    if (sqlite3_prepare_v2(db, sql.toUtf8().constData(), -1, &stmt, nullptr) == SQLITE_OK) {
        sqlite3_bind_text(stmt, 1, pattern.toUtf8().constData(), -1, SQLITE_TRANSIENT);

        while (sqlite3_step(stmt) == SQLITE_ROW) {
            const char *name = reinterpret_cast<const char *>(sqlite3_column_text(stmt, 0));
            const void *encrypted = sqlite3_column_blob(stmt, 1);
            int encLen = sqlite3_column_bytes(stmt, 1);

            if (name && encrypted && encLen > 0) {
                QByteArray encData(static_cast<const char *>(encrypted), encLen);
                QString decrypted = decryptChromiumValue(encData, browser);
                if (!decrypted.isEmpty()) {
                    cookies[QString::fromUtf8(name)] = decrypted;
                }
            }
        }
        sqlite3_finalize(stmt);
    }

    sqlite3_close(db);
    QFile::remove(tmpPath);

    return cookies.isEmpty() ? QMap<QString, QString>() : cookies;
}

QString BrowserCookieExtractor::decryptChromiumValue(const QByteArray &encrypted, const QString &browser)
{
    if (encrypted.startsWith("v10") || encrypted.startsWith("v11")) {
        return decryptAesCookie(encrypted, browser);
    }
    return QString::fromUtf8(encrypted);
}

QString BrowserCookieExtractor::decryptAesCookie(const QByteArray &encrypted, const QString &browser)
{
    if (encrypted.size() < 31)
        return QString();

    const unsigned char *data = reinterpret_cast<const unsigned char *>(encrypted.constData());
    const unsigned char *nonce = data + 3;
    const unsigned char *ciphertext = data + 15;
    int cipherLen = encrypted.size() - 31;
    const unsigned char *tag = data + 15 + cipherLen;

    QByteArray key = getLinuxKey(browser);
    if (key.isEmpty())
        return QString();

    EVP_CIPHER_CTX *ctx = EVP_CIPHER_CTX_new();
    if (!ctx)
        return QString();

    QByteArray plain(cipherLen + 16, 0);
    int outLen = 0, finalLen = 0;

    bool ok = false;
    do {
        if (EVP_DecryptInit_ex(ctx, EVP_aes_128_gcm(), nullptr, nullptr, nullptr) != 1) break;
        if (EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_SET_IVLEN, 12, nullptr) != 1) break;
        if (EVP_DecryptInit_ex(ctx, nullptr, nullptr,
                               reinterpret_cast<const unsigned char *>(key.constData()),
                               nonce) != 1) break;
        if (EVP_DecryptUpdate(ctx, reinterpret_cast<unsigned char *>(plain.data()), &outLen,
                              ciphertext, cipherLen) != 1) break;
        if (EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_SET_TAG, 16, const_cast<unsigned char *>(tag)) != 1) break;
        if (EVP_DecryptFinal_ex(ctx, reinterpret_cast<unsigned char *>(plain.data()) + outLen, &finalLen) != 1) break;
        ok = true;
    } while (false);

    EVP_CIPHER_CTX_free(ctx);

    if (!ok) return QString();
    return QString::fromUtf8(plain.left(outLen + finalLen));
}

QByteArray BrowserCookieExtractor::getLinuxKey(const QString &browser)
{
    QString localStatePath = getLocalStatePath(browser);
    if (!localStatePath.isEmpty() && QFileInfo::exists(localStatePath)) {
        QFile f(localStatePath);
        if (f.open(QIODevice::ReadOnly)) {
            QJsonDocument doc = QJsonDocument::fromJson(f.readAll());
            f.close();
            QJsonObject obj = doc.object();
            QJsonObject osCrypt = obj["os_crypt"].toObject();
            QString encKey = osCrypt["encrypted_key"].toString();
            if (!encKey.isEmpty()) {
                QByteArray decoded = QByteArray::fromBase64(encKey.toUtf8());
                if (decoded.startsWith("DPAPI")) {
                    decoded = decoded.mid(5);
                }
            }
        }
    }

    return deriveKey(browser + " Safe Storage");
}

QByteArray BrowserCookieExtractor::deriveKey(const QString &password)
{
    QByteArray key(16, 0);
    QByteArray pw = password.toUtf8();
    QByteArray salt = QByteArray("saltysalt");

    PKCS5_PBKDF2_HMAC_SHA1(pw.constData(), pw.size(),
                            reinterpret_cast<const unsigned char *>(salt.constData()), salt.size(),
                            1, 16,
                            reinterpret_cast<unsigned char *>(key.data()));
    return key;
}

QString BrowserCookieExtractor::getLocalStatePath(const QString &browser)
{
    if (browser == "chrome")
        return QDir::homePath() + "/.config/google-chrome/Local State";
    else if (browser == "edge")
        return QDir::homePath() + "/.config/microsoft-edge/Local State";
    return QString();
}

QMap<QString, QString> BrowserCookieExtractor::extractFirefoxCookie(const QString &domain)
{
    QString profilesPath = getBrowserPath("firefox");
    if (profilesPath.isEmpty() || !QDir(profilesPath).exists())
        return {};

    QString basePath = QDir::cleanPath(profilesPath + "/..");
    QString profilesIni = basePath + "/profiles.ini";

    QStringList profiles;
    if (QFileInfo::exists(profilesIni)) {
        profiles = parseFirefoxProfiles(profilesIni, profilesPath);
    } else {
        QDir dir(profilesPath);
        for (const QString &entry : dir.entryList(QDir::Dirs | QDir::NoDotAndDotDot)) {
            if (entry.endsWith(".default") || entry.endsWith(".default-release")) {
                profiles.append(profilesPath + "/" + entry);
            }
        }
    }

    for (const QString &profile : profiles) {
        QString cookiesDb = profile + "/cookies.sqlite";
        if (QFileInfo::exists(cookiesDb)) {
            QMap<QString, QString> cookies = readFirefoxCookies(cookiesDb, domain);
            if (!cookies.isEmpty())
                return cookies;
        }
    }

    return {};
}

QStringList BrowserCookieExtractor::parseFirefoxProfiles(const QString &profilesIni, const QString &basePath)
{
    QStringList profiles;
    QFile f(profilesIni);
    if (!f.open(QIODevice::ReadOnly))
        return profiles;

    QString currentSection;
    QString currentPath;
    bool currentIsRelative = false;
    bool inProfile = false;

    QTextStream in(&f);
    while (!in.atEnd()) {
        QString line = in.readLine().trimmed();
        if (line.isEmpty()) continue;

        if (line.startsWith("[") && line.endsWith("]")) {
            if (inProfile && !currentPath.isEmpty()) {
                QString fullPath = currentIsRelative ? basePath + "/" + currentPath : currentPath;
                if (QDir(fullPath).exists())
                    profiles.append(fullPath);
            }
            currentSection = line.mid(1, line.size() - 2);
            inProfile = currentSection.startsWith("Profile");
            currentPath.clear();
            currentIsRelative = false;
        } else if (inProfile) {
            int eq = line.indexOf('=');
            if (eq > 0) {
                QString key = line.left(eq).trimmed().toLower();
                QString val = line.mid(eq + 1).trimmed();
                if (key == "path") currentPath = val;
                else if (key == "isrelative") currentIsRelative = (val == "1");
            }
        }
    }

    if (inProfile && !currentPath.isEmpty()) {
        QString fullPath = currentIsRelative ? basePath + "/" + currentPath : currentPath;
        if (QDir(fullPath).exists())
            profiles.append(fullPath);
    }

    f.close();
    return profiles;
}

QMap<QString, QString> BrowserCookieExtractor::readFirefoxCookies(const QString &dbPath, const QString &domain)
{
    QMap<QString, QString> cookies;

    QTemporaryFile tmpFile;
    if (!tmpFile.open())
        return {};
    QString tmpPath = tmpFile.fileName();
    tmpFile.close();

    if (!QFile::copy(dbPath, tmpPath))
        return {};

    sqlite3 *db = nullptr;
    if (sqlite3_open(tmpPath.toUtf8().constData(), &db) != SQLITE_OK) {
        if (db) sqlite3_close(db);
        QFile::remove(tmpPath);
        return {};
    }

    QString sql = "SELECT name, value FROM moz_cookies WHERE host LIKE ?";
    sqlite3_stmt *stmt = nullptr;
    QString pattern = "%" + domain + "%";

    if (sqlite3_prepare_v2(db, sql.toUtf8().constData(), -1, &stmt, nullptr) == SQLITE_OK) {
        sqlite3_bind_text(stmt, 1, pattern.toUtf8().constData(), -1, SQLITE_TRANSIENT);

        while (sqlite3_step(stmt) == SQLITE_ROW) {
            const char *name = reinterpret_cast<const char *>(sqlite3_column_text(stmt, 0));
            const char *value = reinterpret_cast<const char *>(sqlite3_column_text(stmt, 1));
            if (name && value) {
                cookies[QString::fromUtf8(name)] = QString::fromUtf8(value);
            }
        }
        sqlite3_finalize(stmt);
    }

    sqlite3_close(db);
    QFile::remove(tmpPath);

    return cookies.isEmpty() ? QMap<QString, QString>() : cookies;
}

QMap<QString, QMap<QString, QString>> BrowserCookieExtractor::getAllGithubCookies()
{
    QMap<QString, QMap<QString, QString>> results;

    auto chrome = getGithubCookieFromChrome();
    if (!chrome.isEmpty()) results["chrome"] = chrome;

    auto edge = getGithubCookieFromEdge();
    if (!edge.isEmpty()) results["edge"] = edge;

    auto firefox = getGithubCookieFromFirefox();
    if (!firefox.isEmpty()) results["firefox"] = firefox;

    return results;
}

QStringList BrowserCookieExtractor::listAvailableBrowsers()
{
    QStringList available;
    for (const QString &browser : {"chrome", "edge", "firefox"}) {
        QString path = getBrowserPath(browser);
        if (!path.isEmpty()) {
            if (browser == "firefox") {
                if (QDir(path).exists())
                    available.append(browser);
            } else {
                if (QFileInfo::exists(path))
                    available.append(browser);
            }
        }
    }
    return available;
}

QString BrowserCookieExtractor::formatCookieString(const QMap<QString, QString> &cookies)
{
    QStringList parts;
    for (auto it = cookies.begin(); it != cookies.end(); ++it) {
        parts.append(it.key() + "=" + it.value());
    }
    return parts.join("; ");
}

QMap<QString, QString> getGithubCookieFromBrowser(const QString &browser)
{
    BrowserCookieExtractor extractor;
    if (browser == "auto") {
        auto all = extractor.getAllGithubCookies();
        if (!all.isEmpty())
            return all.first();
    } else if (browser == "chrome") {
        return extractor.getGithubCookieFromChrome();
    } else if (browser == "edge") {
        return extractor.getGithubCookieFromEdge();
    } else if (browser == "firefox") {
        return extractor.getGithubCookieFromFirefox();
    }
    return {};
}

QString getGithubCookieString(const QString &browser)
{
    auto cookies = getGithubCookieFromBrowser(browser);
    if (!cookies.isEmpty())
        return BrowserCookieExtractor::formatCookieString(cookies);
    return QString();
}

QStringList listAvailableBrowsers()
{
    BrowserCookieExtractor extractor;
    return extractor.listAvailableBrowsers();
}
